abi-versions/v0.2.1/README.md

```diff
@@ -0,0 +1,1807 @@
+  - `i32 (const char *) service_name_data`
+  - `i32 (size_t) service_name_size`
+  - `i32 (const char *) method_name_data`
+  - `i32 (size_t) method_name_size`
+  - `i32 (const uint8_t *) serialized_initial_metadata_data`
+  - `i32 (size_t) serialized_initial_metadata_size`
+  - `i32 (uint32_t *) return_stream_id`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Opens gRPC stream with [serialized] `initial_metadata`
+to gRPC `method_name` on gRPC `service_name` on `upstream`.
+
+gRPC messages can be sent on this stream using [`proxy_grpc_send`]
+with `return_stream_id`.
+
+[`proxy_on_grpc_receive`] or [`proxy_on_grpc_close`] will be called
+with `return_stream_id` when the response is received by the host,
+or after the `timeout`.
+
+`return_stream_id` can also be used to cancel outstanding gRPC messages
+using [`proxy_grpc_cancel`] or close this gRPC stream using
+[`proxy_grpc_close`].
+
+Returned `status` value is:
+- `OK` on success.
+- `PARSE_FAILURE` for unknown `upstream`.
+- `INTERNAL_FAILURE` when the host failed to open gRPC stream.
+- `INVALID_MEMORY_ACCESS` when `upstream_data`, `upstream_size`,
+  `service_name_data`, `service_name_size` `method_name_data`,
+  `method_name_size`, `initial_metadata_data`, `initial_metadata_size`
+  and/or `return_stream_id` point to invalid memory address.
+
+
+### `proxy_grpc_send`
+
+* params:
+  - `i32 (uint32_t) stream_id`
+  - `i32 (const uint8_t *) message_data`
+  - `i32 (size_t) message_size`
+  - `i32 (bool) end_stream`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Sends gRPC `message` on the gRPC stream `stream_id` previously
+established using [`proxy_grpc_stream`].
+
+Returned `status` value is:
+- `OK` on success.
+- `BAD_ARGUMENT` for invalid `stream_id`.
+- `NOT_FOUND` for unknown `stream_id`.
+- `INVALID_MEMORY_ACCESS` when `message_data` and/or `message_size`
+  point to invalid memory address.
+
+
+### `proxy_grpc_cancel`
+
+* params:
+  - `i32 (uint32_t) call_or_stream_id`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Cancels `call_or_stream_id` previously started using
+[`proxy_grpc_call`] or [`proxy_grpc_stream`].
+
+Returned `status` value is:
+- `OK` on success.
+- `BAD_ARGUMENT` for invalid `call_or_stream_id`.
+- `NOT_FOUND` for unknown `call_or_stream_id`.
+
+
+### `proxy_grpc_close`
+
+* params:
+  - `i32 (uint32_t) call_or_stream_id`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Closes `call_or_stream_id` previously started using
+[`proxy_grpc_call`] or [`proxy_grpc_stream`].
+
+Returned `status` value is:
+- `OK` on success.
+- `BAD_ARGUMENT` for invalid `call_or_stream_id`.
+- `NOT_FOUND` for unknown `call_or_stream_id`.
+
+
+### `proxy_get_status`
+
+* params:
+  - `i32 (uint32_t *) return_status_code`
+  - `i32 (const char **) return_status_message_data`
+  - `i32 (size_t *) return_status_message_size`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Retrieves gRPC `return_status_code` and `return_status_message`
```

Comment:
Is the `proxy_get_status` hostcall specific to gRPC call statuses? The [GeneralInterface::getStatus](https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/c932df12c9db59d331dcb1020174d5f911f67244/include/proxy-wasm/context_interface.h#L566-L570) method in proxy-wasm-cpp-host that corresponds to this has doc that states:
> Provides the status of the last call into the VM or out of the VM, similar to errno.

Reply:
That was the original plan, but the only thing this currently implements is gRPC status (and HTTP call failure, but that's hardcoded to a single `"reset"` value, so I ignored it).